actions/merkleroots/get_merkleroots_endpoint_test.go

```diff
@@ -0,0 +1,163 @@
+		expectedResponse models.MerkleRootsBHSResponse
+	}{
+		"Get MerkleRoots success no query params": {
+			query: "",
+			expectedResponse: models.MerkleRootsBHSResponse{
+				Content: testabilities.MockedBHSData,
+				Page: models.ExclusiveStartKeyPageInfo{
+					TotalElements:    len(testabilities.MockedBHSData),
+					Size:             len(testabilities.MockedBHSData),
+					LastEvaluatedKey: "",
+				},
+			},
+		},
+		"Get MerkleRoots success with last evaluated key param": {
+			query: "?lastEvaluatedKey=df2b060fa2e5e9c8ed5eaf6a45c13753ec8c63282b2688322eba40cd98ea067a",
+			expectedResponse: models.MerkleRootsBHSResponse{
+				Content: testabilities.MockedBHSData[5:],
+				Page: models.ExclusiveStartKeyPageInfo{
+					TotalElements:    len(testabilities.MockedBHSData),
+					Size:             len(testabilities.MockedBHSData[5:]),
+					LastEvaluatedKey: "",
+				},
+			},
+		},
+	}
+
+	for name, tt := range testCases {
+		t.Run(name, func(t *testing.T) {
+
+			// given
+			expResponseJSON, _ := json.Marshal(tt.expectedResponse)
+			given, then := testabilities.New(t)
+			cleanup := given.StartedSPVWallet()
+			defer cleanup()
+			url := merklerootsURL
+
+			if tt.query != "" {
+				url = url + tt.query
+			}
+
+			// and
+			client := given.HttpClient().ForUser()
+
+			// when
+			res, err := client.R().
+				SetHeader("Content-Type", "application/json").
+				Get(url)
+
+			// then
+			then.Response(res).IsOK().WithJSONf(string(expResponseJSON))
+			require.NoError(t, err)
+		})
+	}
+
+}
+
+func TestGETMerkleRootsFailure(t *testing.T) {
+
+	testCases := map[string]struct {
+		bhsToken         string
+		bhsURL           string
+		expectErr        models.SPVError
+		response         string
+		responseCode     int
+		batchSize        string
+		lastEvaluatedKey string
+	}{
+		"Get MerkleRoots with wrong batch size": {
+			bhsURL:       merklerootsURL,
+			expectErr:    chainerrors.ErrInvalidBatchSize,
+			responseCode: 400,
+			response:     "{\"code\": \"ErrInvalidBatchSize\",\"message\": \"batchSize must be 0 or a positive integer\"}",
+		},
+		"Get MerkleRoots with invalid merkleroot": {
+			bhsURL:       merklerootsURL,
+			expectErr:    chainerrors.ErrMerkleRootNotFound,
+			responseCode: 404,
+			response:     "{\"code\": \"ErrMerkleRootNotFound\",\"message\": \"No block with provided merkleroot was found\"}",
+		},
+		"Get MerkleRoots with stale merkleroot": {
+			bhsURL:       merklerootsURL,
+			expectErr:    chainerrors.ErrMerkleRootNotInLongestChain,
+			responseCode: 409,
+			response:     "{\"code\": \"ErrMerkleRootNotInLC\",\"message\": \"Provided merkleroot is not part of the longest chain\"}",
+		},
+	}
+
+	for name, tt := range testCases {
+		t.Run(name, func(t *testing.T) {
+			// given
+			given, then := testabilities.New(t)
+			cleanup := given.StartedSPVWallet()
+			defer cleanup()
+
+			// and
+			given.BHS().WillRespondForMerkleRoots(tt.responseCode, tt.response)
+			client := given.HttpClient().ForUser()
```

Comment:
you keep repeating this pattern so you could consider extracting these common setup steps into a helper function for better code reuse and readability.
helper function
```go
func setupWalletAndClient(t *testing.T, responseCode *int, response *string) (given *testabilities.Given, cleanup func(), client *resty.Client) {
    given, err := testabilities.New(t)
    require.NoError(t, err)
    
    cleanup = given.StartedSPVWallet()

    if responseCode != nil && response != nil {
        given.BHS().WillRespondForMerkleRoots(*responseCode, *response)
    }

    client = given.HttpClient().ForUser()
    return given, cleanup, client
}
```

then you would have things like this

```suggestion
			// given
            given, cleanup, client := setupWalletAndClient(t, &tt.responseCode, &tt.response)
            defer cleanup()
```

Reply:
I extracted the function and used it in applicable places. LMKWYT? I think the previous way, even though it repeated two times is a little cleaner to read but I'm curious to know what you think